src/codegen/encoding/mod.rs

```diff
@@ -592,35 +997,37 @@ fn finish_array_loop(for_loop: &ForLoop, vartab: &mut Vartable, cfg: &mut Contro
 fn load_struct_member(ty: Type, expr: Expression, field: usize) -> Expression {
     if ty.is_fixed_reference_type() {
         // We should not dereference a struct or fixed array
-        return Expression::StructMember(Loc::Codegen, ty, Box::new(expr), field);
+        return Expression::StructMember(LOC, ty, expr.into(), field);
     }
+    let s = Expression::StructMember(LOC, Type::Ref(ty.clone().into()), expr.into(), field);
+    Expression::Load(LOC, ty, s.into())
+}
+
+/// Get the array length inside a variable.
+fn array_length(arr: &Expression, vartab: &mut Vartable, cfg: &mut ControlFlowGraph) -> Expression {
```

Comment:
Agreed, this gets a bit confusing. However, `get_array_length` loads a subarray and returns the `Builtin::Array` expression on that, where as `array_length` directly uses `Builtin::ArrayLength` and stores that inside an `Expression::Variable`. So they are two distinct helper functions serving a different purpose.

I just decided to nuke `get_array_length` instead, because it used only in one place, and can be refactored to smaller code size (the commit removed 14 lines). What do you think?

Reply:
`get_array_length` works for any of the array dimensions. It is a general implementation. Let's say your array is a `uint8[3][][4][]`. `get_array_length` works for any of the dimensions you want to fetch, this is why the load is necessary. If you are retrieving the last dimension, no load will be added.

As it was only used in one place, there is no problem in removing it. The function you added only considers the last dimension of an array. Its name should reflect what it does. Alternatively, you can just call `get_array_length` for the last dimension. Storing the length on a variable is only necessary if the length is changing throughout the code, which is not the case for the encoding arguments.

You can choose whether to rename `array_length` to something in the lines of `outer_array_length` or to only use `get_array_length`.

Reply:
Doesn't generate `get_array_length` an unnecessary `Subscript` expression if I'm only interested the outer dimension?

>  Storing the length on a variable is only necessary if the length is changing throughout the code
I'm aware of that and this is why I created the helper function, because I needed to this in more than one location.

Reply:
> Doesn't generate `get_array_length` an unnecessary `Subscript` expression if I'm only interested the outer dimension? If this gets optimized away anyways later on in emit, I'm all in for only using this method.

No, there is no subscript generated. Let's say you want dimension 2, and the `dims` vector has a length of 3.

The slice passed to `load_subarray` would be `dims[3..3]`, which has a length of zero.
https://github.com/hyperledger/solang/blob/468e4d6a59794e0ac4a74001986f600bec0a6d68/src/codegen/encoding/mod.rs#L385

Later, in function `load_subarray` the loop does not execute, because the slice has a length of zero. This means the function only returns the array itself without any subscript.

https://github.com/hyperledger/solang/blob/468e4d6a59794e0ac4a74001986f600bec0a6d68/src/codegen/encoding/mod.rs#L462

Reply:
Sorry I overlooked this. Makes sense, thanks